Random documentation (#356)
BrightSpots/rcv: src/main/resources/network/brightspots/rcv/config_file_documentation.txt

    set of configuration parameters that specify the tabulation rules
    the "rules" section contains the following parameters:

      "randomSeed" required when tiebreakMode is random, previousRoundCountsThenRandom, or generatePermutation
        the integer seed for the application's pseudorandom number generator

>         value: [0..2147483647]

Reviewer: It would be more robust and safer to allow larger seeds, ala the 48 bits actually used by Java, or the 20 decimal digits typical for election auditing approaches.

Reviewer 2: Sure, we can do this.

        if not supplied: null

      "tiebreakMode" required
        how the program should decide which candidate to eliminate when multiple candidates are tied for last place
        or which candidate to elect first when:

>           1) electionWinnerMode is set to multiSeatAllowOnlyOneWinnerPerRound, and

Reviewer: Would this also apply if it is the final round even if multiSeatAllowOnlyOneWinnerPerRound is not the current mode?

Reviewer 2: When multiSeatAllowOnlyOneWinnerPerRound isn't enabled and multiple candidates exceed the threshold in the same round, we consider them to be elected simultaneously. Their surplus redistributions happen after they've been all been elected.

          2) multiple candidates exceed the winning threshold in the same round, and
          3) at least two of those candidates are tied for the highest vote total in that round
        value: "random" | "interactive" | "previousRoundCountsThenRandom" | "previousRoundCountsThenInteractive" | "usePermutationInConfig" | "generatePermutation"

>         we use Java 12 java.util.random for randomness in our tiebreak implementations

Reviewer: Saying "12" makes it seem specific to that version, which we all dearly hope it will never be.
This phrasing hopefully makes it seem more stable:
> "we use java.util.random for randomness in our tiebreak implementations. Compatible
 methods exist for other languages, e.g. https://pypi.org/project/java-random/

Author: The .jar file we create is compiled against Java 12, so this statement is accurate.  If one wanted to reproduce our results you would need to do the same, no?  It's not to say we're somehow tied to this implementation in future releases.

Reviewer: You bring up an important topic: what sort of reproducability are we trying to achieve?
As I just noted at https://github.com/BrightSpots/rcv/issues/43#issuecomment-509695201
I think the documentation, code and comments should all help clarify and implement the goal of defining and producing a stable standard definition of everything necessary for folks who want to write compatible software that comes up with exactly the same results as this VVSG reference implementation.

So I'm looking for a more stable definition of the PRNG in use than saying it is what Java ships in version 12. I think it has actually be stable for a long time and will remain stable, and we should frame comments to reflect that, and to help those who are writing compatible software in other languages.

Reviewer 3: @nealmcb changed to your wording in PR #370.

Author: Gotcha.  That is the notion of reproducibility I tried to facilitate in this PR so I think we are on the same page.  Is there a library you have in mind?  Practically speaking, for this release we are not going to include any code changes like this, but we could further enhance the documentation, and potentially upgrade our implementation in the future.

    set of configuration parameters that specify the tabulation rules
    the "rules" section contains the following parameters:

      "randomSeed" required when tiebreakMode is random, previousRoundCountsThenRandom, or generatePermutation
        the integer seed for the application's pseudorandom number generator
        value: [0..2147483647]
        if not supplied: null

      "tiebreakMode" required
        how the program should decide which candidate to eliminate when multiple candidates are tied for last place
        or which candidate to elect first when:

>           electionWinnerMode is set to multiSeatAllowOnlyOneWinnerPerRound and

Reviewer 2: ```suggestion
          1) electionWinnerMode is set to multiSeatAllowOnlyOneWinnerPerRound, and
```

>           multiple candidates exceed the winning threshold and

Reviewer 2: ```suggestion
          2) multiple candidates exceed the winning threshold in the same round, and
```

>           also tie for the highest vote total in the same round

Reviewer 2: ```suggestion
          3) at least two of thoes candidates are tied for the highest vote total in that round 
```

        we use Java 12 java.util.random in our tiebreak implementations as follows:

        "random"
          on config load a java.util.Random object is created using the randomSeed value specified in the input config file:
          random = new Random(randomSeed);
          during tabulation, in the event of a tie at the end of a round:
          the list of tied candidates is sorted alphabetically
          a randomDouble is created using the random object:
            double randomDouble = random.nextDouble();
            the randomDouble is mapped to one of the tied candidates in the list:
            int randomCandidateIndex = (int) Math.floor(randomDouble * (double) tiedCandidates.size());
          the selected candidate will be the winner or loser for that round

        "interactive":
          the user is presented with a list of tied candidates
          the user will input their selection manually

        "previousRoundCountsThenRandom"
          the tied candidate with the highest vote total in the previous round is selected

>           if there is a tie for the vote count in the previous round as well, a candidate is selected from the still tying candidates as described under tiebreakMode "random"

Reviewer 2: I actually haven't implemented this part properly. If it has to fall back to random or interactive, it picks from the full set of tied candidates, not just the ones who were still alive at the end of the previousRoundCounts logic. I'll send a PR to fix this shortly.
